ssl/s3_lib.c

```diff
@@ -5036,7 +5036,14 @@ int ssl_derive(SSL_CONNECTION *s, EVP_PKEY *privkey, EVP_PKEY *pubkey, int gense
     }
 
     if (EVP_PKEY_derive(pctx, pms, &pmslen) <= 0) {
-        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        /*
+         * the public key probably was a weak key
+         */
+        if (ERR_GET_REASON(ERR_peek_last_error()) == ERR_LIB_EC) {
+            SSLfatal(s, SSL_AD_INSUFFICIENT_SECURITY, SSL_R_BAD_ECPOINT);
+        } else {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        }
```

Comment:
I would not check the reason code at all and used  SSL_AD_ILLEGAL_PARAMETER instead as suggested in #27531. The returned reason code should be SSL_R_BAD_KEY_SHARE.

Reply:
The `EVP_PKEY_derive` can fail on many different reasons; would it be good to report all of them as ILLEGAL_PARAMETER?

Reply:
It is IMO OK. The alert is just an alert. This is the most common case of the failure.